File: reference/CodeExamples/dotNet/VisualBasicExamples/Code/troubleshooting/restore/TryToReadSingleObjects.vb
Imports Db4objects.Db4o

Namespace Db4oDoc.Code.Troubleshooting.Restore
    Public Class TryToReadSingleObjects
        Private Const DatabaseFileName As String = "database.db4o"

        Public Shared Sub Main(args As String())
            StoreExampleObjects()
            Using container As IObjectContainer = Db4oEmbedded.OpenFile(DatabaseFileName)
                ' #example: Try to read the intact objects
                Dim idsOfPersons As Long() = container.Ext().StoredClass(GetType(Person)).GetIDs()
                For Each id As Long In idsOfPersons
                    Try
                        Dim person = DirectCast(container.Ext().GetByID(id), Person)
                        container.Ext().Activate(person, 1)
                        ' store the person to another database
                        Console.Out.WriteLine("This object is ok {0}", person)
                    Catch e As Exception
                        Console.Out.WriteLine("We couldn't read the object with the id {0} anymore." & " It is lost", id)
                        Console.Out.WriteLine(e)
                    End Try
                Next
                ' #end example
            End Using
        End Sub

        Private Shared Sub StoreExampleObjects()
            Using container As IObjectContainer = Db4oEmbedded.OpenFile(DatabaseFileName)
                For i As Integer = 0 To 99
                    container.Store(New Person("Fun" & i))
                Next
            End Using
        End Sub
    End Class


    Friend Class Person
        Private ReadOnly m_name As String

        Public Sub New(name As String)
            Me.m_name = name
        End Sub

        Public ReadOnly Property Name() As String
            Get
                Return m_name
            End Get
        End Property
    End Class
End Namespace

File: reference/CodeExamples/dotNet/CSharpExamples/Code/troubleshooting/restore/TryToReadSingleObjects.cs
using System;
using Db4objects.Db4o;

namespace Db4oDoc.Code.Troubleshooting.Restore
{
    public class TryToReadSingleObjects
    {
        private const string DatabaseFileName = "database.db4o";

        public static void Main(string[] args)
        {
            StoreExampleObjects();
            using (IObjectContainer container = Db4oEmbedded.OpenFile(DatabaseFileName))
            {
                // #example: Try to read the intact objects
                long[] idsOfPersons = container.Ext().StoredClass(typeof (Person)).GetIDs();
                foreach (long id in idsOfPersons)
                {
                    try
                    {
                        var person = (Person) container.Ext().GetByID(id);
                        container.Ext().Activate(person, 1);
                        // store the person to another database
                        Console.Out.WriteLine("This object is ok {0}", person);
                    }
                    catch (Exception e)
                    {
                        Console.Out.WriteLine("We couldn't read the object with the id {0} anymore." +
                                              " It is lost", id);
                        Console.Out.WriteLine(e);
                    }
                }
                // #end example
            }
        }

        private static void StoreExampleObjects()
        {
            using (IObjectContainer container = Db4oEmbedded.OpenFile(DatabaseFileName))
            {
                for (int i = 0; i < 100; i++)
                {
                    container.Store(new Person("Fun" + i));
                }
            }
        }
    }


    internal class Person
    {
        private readonly string name;

        public Person(string name)
        {
            this.name = name;
        }

        public string Name
        {
            get { return name; }
        }
    }
}